### DISCUSSION TITLE
Spacing vs Spacingd transforms

### DISCUSSION THREAD

**Discussion by natwille1** [Project-MONAI/tutorials#691] [Q&A]
Hi,

I am seeing different results when comparing the output of the Spacing vs Spacingd transforms. 

I have a dataset of CT volumes and their corresponding masks. I create and apply the transforms as follows:

```
keys=['image', 'label']
loader = transforms.LoadImaged(keys=keys)
addchanneld = transforms.AddChanneld(keys=keys)
addchannel = transforms.AddChannel()
samplerd = transforms.Spacingd(keys=keys, 
                               pixdim=(1.2, 1.2, 1.2), mode=('bilinear', 'nearest')
                               )
sampler = transforms.Spacing((1.2, 1.2, 1.2))

data_loaded = loader(data_dicts[0])
data_chan = addchanneld(data_loaded)
data_sampled_dict  = samplerd(data_chan)

image = data_chan['image']
label = data_chan['label']

image_sampled = sampler(image)
label_sampled = sampler(label)

print(image.shape, label.shape)
print(image_sampled[0].shape, label_sampled[0].shape)
print(data_sampled_dict['image'].shape, data_sampled_dict['label'].shape)
```

And I get the following output:

```
(1, 512, 512, 300) (1, 512, 512, 300)
(1, 427, 427, 250) (1, 427, 427, 250)
(1, 209, 209, 157) (1, 427, 427, 250)
```

as you can see, the processed image and processed label shapes are the same when using the Spacing transform (second output line) but they differ when using the dictionary transform (last line of output). 

Here are the metadata dictionaries of the image and label that are passed to the Spacingd dictionary transform:

```
IMAGE:
'spacing': array([0.48828101, 0.48828101, 0.625     ]),
  'original_affine': array([[  -0.48828101,    0.        ,    0.        ,    6.1       ],
         [   0.        ,   -0.48828101,    0.        ,  104.2       ],
         [   0.        ,    0.        ,    0.625     , -109.375     ],
         [   0.        ,    0.        ,    0.        ,    1.        ]]),
  'affine': array([[  -1.2  ,    0.   ,    0.   ,    6.1  ],
         [   0.   ,   -1.2  ,    0.   ,  104.2  ],
         [   0.   ,    0.   ,    1.2  , -109.375],
         [   0.   ,    0.   ,    0.   ,    1.   ]], dtype=float32),

LABEL:
 'label_meta_dict': {'spacing': array([0.48828101, 0.48828101, 0.625     ]),
  'original_affine': array([[  -0.48828101,    0.        ,    0.        ,    6.1       ],
         [   0.        ,   -0.48828101,    0.        ,  104.2       ],
         [   0.        ,    0.        ,    0.625     , -109.375     ],
         [   0.        ,    0.        ,    0.        ,    1.        ]]),
  'affine': array([[-1.2,  0. ,  0. ,  0. ],
         [ 0. , -1.2,  0. ,  0. ],
         [ 0. ,  0. ,  1.2,  0. ],
         [ 0. ,  0. ,  0. ,  1. ]], dtype=float32)
```

Why do I get different outputs depending on whether I use the dictionary-based transform? Thanks!

**Comment by eyalhizmi** [comment]
In Spacing label you need to add  mode= 'nearest' . like this : sampler(load, mode= 'nearest')
the default mode is bilinear and this is the reason you get different output. 
Notice that in spacingd the mode written in sequence of string, each element corresponds to a key in keys.

**Comment by JayaBotsa** [comment]
Hi all,

I am trying to use the Spacing to resample the mri data but when I am using the affine attribute on it I am getting an error:                                  
arg `affine` is deprecated, the affine of MetaTensor in data_array has higher priority.

**Reply by KumoLiu** [reply]
Yes, `affine` is deprecated since 0.9, the original affine information will be get from the MetaTensor.
https://github.com/Project-MONAI/MONAI/blob/c33f1ba588ee00229a309000e888f9817b4f1934/monai/transforms/spatial/array.py#L496
